File: prolog/04-control-flow/recursion.pl
% recursion.pl - Recursion replaces loops

% Countdown: a "loop" expressed as recursion
countdown(0) :- writeln('Liftoff!').
countdown(N) :-
    N > 0,
    writeln(N),
    N1 is N - 1,
    countdown(N1).

% Factorial: accumulate a result through the recursive calls
factorial(0, 1).
factorial(N, F) :-
    N > 0,
    N1 is N - 1,
    factorial(N1, F1),
    F is N * F1.

% Sum a list by walking its head and tail
sum_list_recursive([], 0).
sum_list_recursive([H|T], Sum) :-
    sum_list_recursive(T, Rest),
    Sum is H + Rest.

:- writeln('Countdown from 3:'),
   countdown(3),
   factorial(5, F),
   format('5! = ~w~n', [F]),
   sum_list_recursive([10, 20, 30, 40], S),
   format('Sum of [10,20,30,40] = ~w~n', [S]),
   halt.

File: erlang/05-functions/recursion.erl
#!/usr/bin/env escript
%% Recursion replaces loops in Erlang

main(_) ->
    %% Body-recursive factorial
    io:format("factorial(6) = ~p~n", [factorial(6)]),

    %% Recursing over a list with [Head | Tail]
    io:format("sum([1,2,3,4,5]) = ~p~n", [sum([1, 2, 3, 4, 5])]),
    io:format("len([a,b,c,d]) = ~p~n", [len([a, b, c, d])]),

    %% Tail recursion: an accumulator avoids growing the stack
    io:format("factorial_tail(6) = ~p~n", [factorial_tail(6)]),

    io:format("count_down(5): ", []),
    count_down(5),
    io:format("~n", []).

%% Body-recursive: the multiply happens after the recursive call returns
factorial(0)            -> 1;
factorial(N) when N > 0 -> N * factorial(N - 1).

%% Walk a list: base case is the empty list, then handle head + tail
sum([])    -> 0;
sum([H|T]) -> H + sum(T).

len([])    -> 0;
len([_|T]) -> 1 + len(T).

%% Tail-recursive: a public function delegates to a private helper/2
factorial_tail(N) -> factorial_tail(N, 1).

factorial_tail(0, Acc)            -> Acc;
factorial_tail(N, Acc) when N > 0 -> factorial_tail(N - 1, N * Acc).

%% Recursion can also drive side effects, like printing
count_down(0) -> io:format("liftoff!");
count_down(N) when N > 0 ->
    io:format("~p ", [N]),
    count_down(N - 1).
